pkg/volume/util/atomic_writer.go

```diff
@@ -446,13 +436,18 @@ func (w *AtomicWriter) createUserVisibleFiles(payload map[string]FileProjection)
 // removeUserVisiblePaths removes the set of paths from the user-visible
 // portion of the writer's target directory.
 func (w *AtomicWriter) removeUserVisiblePaths(paths sets.String) error {
-	orderedPaths := paths.List()
-	for ii := len(orderedPaths) - 1; ii >= 0; ii-- {
-		if err := os.Remove(path.Join(w.targetDir, orderedPaths[ii])); err != nil {
-			glog.Errorf("%s: error pruning old user-visible path %s: %v", w.logContext, orderedPaths[ii], err)
-			return err
+	ps := string(os.PathSeparator)
+	var lasterr error
+	for p := range paths {
+		// only remove symlinks from the volume root directory (i.e. items that don't contain '/')
+		if strings.Contains(p, ps) {
+			continue
+		}
+		if err := os.Remove(path.Join(w.targetDir, p)); err != nil {
+			glog.Errorf("%s: error pruning old user-visible path %s: %v", w.logContext, p, err)
+			lasterr = err
```

Comment:
Removal failures are not immediately fatal now.  We could lose errors.  Just pointing it out.

Reply:
I did this on purpose, but I don't know that it's better. My thinking was "If we have a problem removing 1 file, isn't it better to clean up as much of the rest of it as possible?" but I'm not sure what the answer is. The other problem with the new approach is that any errors but the last one will not be seen by the caller (but will be in the error log). I'm happy with either behavior (old/new).